### ISSUE TITLE
Alternates for Dynamic Parts Not Added to Shape Metadata

### ISSUE THREAD

**Issue by aodl** [OrchardCMS/OrchardCore#2363]
Content [part templates](https://orchardcore.readthedocs.io/en/latest/OrchardCore.Modules/OrchardCore.Templates/README/#content-part-templates) are not available to dynamically defined parts (a dynamically defined part doesn't make use of the `ContentPartDisplayDriver<TPart>`, which is currently responsible for adding the necessary alternates). 

It would be useful (and less misleading for new comers) if [these templates](https://orchardcore.readthedocs.io/en/latest/OrchardCore.Modules/OrchardCore.Templates/README/#content-part-templates) were available by default for dynamically defined parts as well as strongly typed parts (those with corresponding C# classes).

Do others agree that this is first a problem? and that it is worth solving?

### Possible Approach

The solution seems to require a default ContentPartDisplayDriver which operates on the basic `ContentPart` object (as opposed to ContentPart subclasses). `ContentItemDisplayCoordinator.BuildDisplayAsync` is responsible for looping through the part drivers that have been registered, and calling every driver's `BuildDisplayAsync` for every part. 
```
    foreach (var displayDriver in _partDisplayDrivers)
    {
        try
        {
            var result = await displayDriver.BuildDisplayAsync(part, contentTypePartDefinition, context);
            if (result != null)
            {
                await result.ApplyAsync(context);
            }
       ...
```

The part drivers' `BuildDisplayAsync` methods are responsible for checking if they should return early (because the part argument passed in is not of the correct type). i.e. the first few lines look like this 
```
    var part = contentPart as TPart;
    if (part == null)
    {
        return Task.FromResult<IDisplayResult>(null);
    }
```

#### Implementing a Solution

If a default ContentPartDisplayDriver would be implemented, inheriting the existing `ContentPartDisplayDriver<TPart>` (where TPart : ContentPart), it could override the BuildDisplayAsync method and perform a more strict type check before returning early (else calling the base implementation) i.e.
```
    if (part.GetType() != typeof(ContentPart))
    {
        return Task.FromResult<IDisplayResult>(null);
    }
```
This way this default driver would only operate on parts that are not subclasses of ContentPart (indicating that they are dynamically defined in admin UI because they do not have their own ContentPart subclass defined in C#).

I'm not sure if other changes would be required to support this, and I haven't tried it myself yet. It's just a theory before I shoot off for work.

cc @sebastienros, @jtkech (as per our brief discussion on gitter)

**Comment by aodl** [comment]
I'd enjoy working on this if that's okay. What's the protocol for being added as a collaborator for this repository?

**Comment by sfmskywalker** [comment]
That would be fantastic. Here is some documentation on contributing: http://docs.orchardproject.net/en/latest/Documentation/Contributing-patches/#Branches

You already opened an issue, so next step is to fork the repo, do the work and submit a PR.

**Comment by aodl** [comment]
Great thanks. It sounds like its normal in this repo for issues to be 'unassigned' while people work on them. I just wanted to avoid someone else doubling up on the effort because there's no official assignee.

Thanks for the link, I'll get started

**Comment by aodl** [comment]
Btw @sfmskywalker thank you for writing [this hugely informative article](http://www.ideliverable.com/blog/a-closer-look-at-content-types-drivers-shapes-and-placement) about shapes and the rendering process. It really helped me get my head around it :)

**Comment by sfmskywalker** [comment]
You’re welcome 😊

**Comment by mikecasas** [comment]
@sfmskywalker Just an FYI, but once on the link for contributing, the repo listed isn't for Orchard Core.

**Comment by sfmskywalker** [comment]
You’re right. Perhaps this url is more accurate: https://github.com/OrchardCMS/OrchardCore/wiki/Engineering-Guidelines

**Comment by mikecasas** [comment]
@sfmskywalker Also, want to agree with @aodl , your post has been extremely helpful! Thanks!

http://www.ideliverable.com/blog/a-closer-look-at-content-types-drivers-shapes-and-placement

**Comment by sebastienros** [comment]
Let me rephrase the issue to be sure I understand:

- Create a new part `Address`
- Create a type `Customer` and add the `Title` and `Address` parts
- Create a template named `Address`
- Display a customer content item, the custom template is not used

**Comment by aodl** [comment]
Yes, the content part templates are ignored for parts that are created via the admin dashboard. This is only a problem for dynamically defined parts (created via admin dashboard), because they don't have a dedicated content part driver. That is what this PR is about - providing a default content part driver for all dynamically defined parts, ensuring that `ContentPartDisplayDriver<TPart>`'s `Factory` method gets called for those parts (which is where the standard content part alternates are added to the shapes)
